### PR TITLE
8341094: Clean up relax_verify in ClassFileParser

### PR DESCRIPTION
**Pull request by coleenp**
Removed ClassFileParser::_relax_verify and cleaned up setting need_verify in the ClassFileStream code and ClassFileParser.  Wrote up how I found control to flow to this setting with -Xverify:all/none/remote settings, which become BytecodeVerificationRemote/Local in the CR comments.

Tested with tier1-4.  Valid class, field and method name testing is done through the JCK and there are CDS tests that test verification in runtime/cds/appcds/VerifierTest.java.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341094](https://bugs.openjdk.org/browse/JDK-8341094): Clean up relax_verify in ClassFileParser (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21954/head:pull/21954` \
`$ git checkout pull/21954`

Update a local copy of the PR: \
`$ git checkout pull/21954` \
`$ git pull https://git.openjdk.org/jdk.git pull/21954/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21954`

View PR using the GUI difftool: \
`$ git pr show -t 21954`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21954.diff">https://git.openjdk.org/jdk/pull/21954.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21954#issuecomment-2462276390)
</details>

### CODE REVIEW

**Comment by coleenp** [comment]
I don't know why I didn't get email for this PR.

**Comment by iklam** [comment]
I am honestly very confused by this code:

- We always create ClassFileStream with _need_verify = true
- Sometimes we set it back to false in ClassFileParser::ClassFileParser()

```
  _need_verify = Verifier::should_verify_for(_loader_data->class_loader());
  stream->set_check_truncation(_need_verify);
```

Can we instead set `ClassFileStream::_need_verify`once and for all inside the `ClassFileStream` constructor, and use it as a constant afterwards (and propagate to ClassFilePaser, etc)? This can be done by passing a `ClassLoader` into the `ClassFileStream` constructor and call `Verifier::should_verify_for` in there.

Then, you can change `_check_truncation` back to `_need_verify`.

`ClassFileStream::verify` can be removed as it doesn't seem to serve any purpose anymore.

File: src/hotspot/share/classfile/verifier.hpp [hunk 1]
```diff
@@ -52,7 +52,7 @@ class Verifier : AllStatic {
   // Return false if the class is loaded by the bootstrap loader,
   // or if defineClass was called requesting skipping verification
   // -Xverify:all overrides this value
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/verifier.hpp:54]: This comment doesn't describe the new operation of this method.

File: src/hotspot/share/classfile/classFileStream.hpp [hunk 2]
```diff
@@ -52,12 +52,12 @@ class ClassFileStream: public ResourceObj {
   const char* clone_source() const;
 
  public:
-  static const bool verify;
+  constexpr static bool verify = true;
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classFileStream.hpp:55]: Where, and how, is this used?

**Review by dholmes-ora — COMMENTED** [review]
I won't pretend that I could follow your analysis in the full. At the end of the day we can relax verification (of all kinds) for trusted classes in the released product, but we should have a way to enforce all verification checks on trusted classes during development in case we introduce a bug and start producing invalid class files. I'm not sure that this latter part is actually in place unless "verify all" is explicitly enabled. Of course we will discover bad trusted class files one way or another, but enabling verification would make it more obvious as to the problem.

Doing additional checks on trusted classes is okay provided it doesn't affect startup or general performance - if verification (and related checks) were free we'd always do them, so there has to be a tipping point.

Thanks

File: src/hotspot/share/classfile/classFileStream.hpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/classFileStream.hpp:55]: It's passed into the ClassFileStream constructor with default value true.

**Comment by coleenp** [comment]
Ioi, I like your suggestion.  That would make it even clearer.  We could set the value of _need_verify once in ClassFileStream constructor, and propagate the value of _need_verify to ClassFileParser.  ClassFileParser has a lot of checks for _need_verify so it's still better for ClassFileParser to have one.

David, there are three different concepts of "trusted" in this code which caused the duplication that I'm trying to resolve.  One is for package access which includes boot/platform/system class loader.
Two is for this relax_verify to check the names of class/fields/methods which includes boot/platform class loader.
Three is for checking stream truncation, doing classfile format checks and running the Verifier itself which only includes the boot class loader.

My change removes #2 since checking names is a fast enough operation (doesn't call java unless < JDK 1.5), and most , and maybe all platform class loading is called with ClassFileStream::_need_verify == true anyway, so this code checked these names anyway.

Further, I'm trying to only check for when we want verification #3 in only one place as it was set and reset between ClassFileStream and ClassFileParser, and the same code is called again effectively ignoring the value of should_verify_class value passed into the Verifier but getting the same result as the last call.

**Comment by coleenp** [comment]
Ioi, your suggestion of passing class_loader to stream creation is really a cumbersome change.  The Stream's version of need_verify is an internal flag used for determining whether we need to check for truncation and should be based on the ClassFileParser's call when we're about to parse the stream.  But I did some cleanup to further unexpose this field.

File: src/hotspot/share/classfile/classFileStream.hpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/classFileStream.hpp]: I removed this and fixed all the constructor calls.

File: src/hotspot/share/classfile/verifier.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/classfile/verifier.hpp:54]: You're right.  There was an old change that made defineClass explicitly disable verification.  That code is gone now.

**Review by dholmes-ora — COMMENTED** [review]
I like the update so that CFS defaults to needing verification unless CFP sets it otherwise.

**Comment by iklam** [comment]
Why do we need `ClassFileStream::_need_verify`? Is it for performance? If so, we are making a trade off here -- boot classes will be marginally faster, but non-boot classes will be marginally slower. Assuming that large apps load many more non-boot classes than boot classes, this seems counter productive.

Maybe we should get rid of `ClassFileStream::_need_verify` altogether?

**Comment by dholmes-ora** [comment]
> Maybe we should get rid of ClassFileStream::_need_verify altogether?

Get rid of it in what way? Implicitly always true, or implicitly always false? Or?

**Comment by iklam** [comment]
> > Maybe we should get rid of ClassFileStream::_need_verify altogether?
> 
> Get rid of it in what way? Implicitly always true, or implicitly always false? Or?

Always true. E.g.,

```
  u2 get_u2(TRAPS) const {
-   if (_need_verify) {
      guarantee_more(2, CHECK_0);
-   } else {
-     assert(2 <= _buffer_end - _current, "buffer overflow");
-   }
    return get_u2_fast();
  }
```

**Comment by dholmes-ora** [comment]
> Always true.

We need to check performance here, particularly startup, as that is what would be most affected by always verifying boot classes.

**Comment by coleenp** [comment]
Yes, the classFileStream::_need_verify was added for performance.

**Review by iklam — APPROVED** [review]
LGTM. This simplifies the code. There doesn't seem to be security risks as this PR only strengths the checks.

**Review by dholmes-ora — APPROVED** [review]
If the changes are now stable then I am fine with them.

Thanks

**Pull request closed without merging**